src/Workspaces/Core/Portable/Workspace/Solution/DocumentState_LinkedFileReuse.cs

```diff
@@ -0,0 +1,208 @@
+                siblingTextSource,
+                LoadTextOptions,
+                newTreeSource);
+
+            static AsyncLazy<TreeAndVersion> GetReuseTreeSource(
+                string filePath,
+                HostLanguageServices languageServices,
+                LoadTextOptions loadTextOptions,
+                ParseOptions parseOptions,
+                ValueSource<TreeAndVersion> treeSource,
+                ITextAndVersionSource siblingTextSource,
+                ValueSource<TreeAndVersion> siblingTreeSource)
+            {
+                return new AsyncLazy<TreeAndVersion>(
+                    cancellationToken => TryReuseSiblingTreeAsync(filePath, languageServices, loadTextOptions, parseOptions, treeSource, siblingTextSource, siblingTreeSource, cancellationToken),
+                    cancellationToken => TryReuseSiblingTree(filePath, languageServices, loadTextOptions, parseOptions, treeSource, siblingTextSource, siblingTreeSource, cancellationToken),
+                    cacheResult: true);
+            }
+
+            static bool TryReuseSiblingRoot(
+                string filePath,
+                HostLanguageServices languageServices,
+                LoadTextOptions loadTextOptions,
+                ParseOptions parseOptions,
+                SyntaxNode siblingRoot,
+                VersionStamp siblingVersion,
+                [NotNullWhen(true)] out TreeAndVersion? newTreeAndVersion)
+            {
+                var siblingTree = siblingRoot.SyntaxTree;
+
+                // Look for things that disqualify us from being able to use our sibling's root.
+                if (!CanReuseSiblingRoot())
+                {
+                    newTreeAndVersion = null;
+                    return false;
+                }
+
+                var treeFactory = languageServices.GetRequiredService<ISyntaxTreeFactoryService>();
+
+                var newTree = treeFactory.CreateSyntaxTree(
+                    filePath,
+                    parseOptions,
+                    siblingTree.Encoding,
+                    loadTextOptions.ChecksumAlgorithm,
+                    siblingRoot);
+
+                newTreeAndVersion = new TreeAndVersion(newTree, siblingVersion);
+                return true;
+
+                bool CanReuseSiblingRoot()
+                {
+                    s_tryReuseSyntaxTree++;
+                    var siblingParseOptions = siblingTree.Options;
+
+                    var ppSymbolsNames1 = parseOptions.PreprocessorSymbolNames;
+                    var ppSymbolsNames2 = siblingParseOptions.PreprocessorSymbolNames;
+
+                    // If both documents have the same preprocessor directives defined, then they'll always produce the
+                    // same trees.  So we can trivially reuse the tree from one for the other.
+                    if (ppSymbolsNames1.SetEquals(ppSymbolsNames2))
+                    {
+                        s_couldReuseBecauseOfEqualPPNames++;
```

Comment:
Interlocked.Increment? I recognize this is just being used for the benchmark code best to have the good practice?

Reply:
sure.

Reply:
done.